enterprise/internal/batches/service/workspace_resolver.go

```diff
@@ -153,37 +153,59 @@ func (wr *workspaceResolver) ResolveWorkspacesForBatchSpec(ctx context.Context,
 }
 
 func (wr *workspaceResolver) determineRepositories(ctx context.Context, batchSpec *batcheslib.BatchSpec) ([]*RepoRevision, error) {
-	seen := map[api.RepoID]*RepoRevision{}
+	// We have to track each repo that is matched, including whether the
+	// revisions should be overridden or not by future rules.
+	type repo struct {
+		overridable bool
+		revisions   map[string]*RepoRevision
+	}
+	repos := map[api.RepoID]*repo{}
 
 	var errs error
 	// TODO: this could be trivially parallelised in the future.
 	for _, on := range batchSpec.On {
-		repos, err := wr.resolveRepositoriesOn(ctx, &on)
+		revs, overridable, err := wr.resolveRepositoriesOn(ctx, &on)
 		if err != nil {
 			errs = multierror.Append(errs, errors.Wrapf(err, "resolving %q", on.String()))
 			continue
 		}
 
-		for _, repo := range repos {
+		for _, rev := range revs {
 			// Skip repos where no branch exists.
-			if !repo.HasBranch() {
+			if !rev.HasBranch() {
 				continue
 			}
 
-			if other, ok := seen[repo.Repo.ID]; !ok {
-				seen[repo.Repo.ID] = repo
+			if seenRepo, ok := repos[rev.Repo.ID]; ok {
+				// We've seen this repo previously, so we now need to ascertain
+				// whether we're overriding the previous revisions, or merging
+				// into them.
+				if seenRepo.overridable {
+					seenRepo.revisions = map[string]*RepoRevision{
+						rev.Branch: rev,
+					}
+				} else {
+					seenRepo.revisions[rev.Branch] = rev
+				}
+				seenRepo.overridable = overridable
 			} else {
-				// If we've already seen this repository, we overwrite the
-				// Commit/Branch fields with the latest value we have
-				other.Commit = repo.Commit
-				other.Branch = repo.Branch
+				// This is a new repo, so let's set up the state we need to
+				// track.
+				repos[rev.Repo.ID] = &repo{
+					overridable: overridable,
+					revisions: map[string]*RepoRevision{
+						rev.Branch: rev,
+					},
+				}
 			}
 		}
 	}
 
-	repoRevs := make([]*RepoRevision, 0, len(seen))
-	for _, rr := range seen {
-		repoRevs = append(repoRevs, rr)
+	repoRevs := []*RepoRevision{}
```

Comment:
We could optimize for the standard case here and do 

```suggestion
	repoRevs := make([]*RepoRevision, 0, len(repos))
```

Reply:
Now moot with the aggregator refactor.